tendermint/src/vote.rs

```diff
@@ -28,26 +42,69 @@ pub struct Vote {
 
     /// Round
     #[serde(with = "serializers::from_str")]
-    pub round: u32,
+    pub round: block::Round,
 
     /// Block ID
     #[serde(deserialize_with = "serializers::parse_non_empty_block_id")]
     pub block_id: Option<block::Id>,
 
     /// Timestamp
-    pub timestamp: Time,
+    pub timestamp: Option<Time>,
 
     /// Validator address
     pub validator_address: account::Id,
 
     /// Validator index
     #[serde(with = "serializers::from_str")]
-    pub validator_index: u16,
+    pub validator_index: ValidatorIndex,
 
     /// Signature
     pub signature: Signature,
 }
 
+impl DomainType<RawVote> for Vote {}
+
+impl TryFrom<RawVote> for Vote {
+    type Error = Error;
+
+    fn try_from(value: RawVote) -> Result<Self, Self::Error> {
+        if value.timestamp.is_none() {
+            return Err(NoTimestamp.into());
+        }
+        Ok(Vote {
+            vote_type: value.r#type.try_into()?,
+            height: value.height.try_into()?,
+            round: value.round.try_into()?,
+            block_id: match value.block_id {
+                None => None,
+                Some(b) => Some(b.try_into()?),
+            },
+            timestamp: match value.timestamp {
+                None => None,
+                Some(t) => Some(t.try_into()?),
+            },
```

Comment:
As above, these two expressions could be simplified using [`Option::transpose`](https://doc.rust-lang.org/beta/std/option/enum.Option.html#method.transpose).

Reply:
Covered in fixes #3 commit.